cluster/query_executor.go

```diff
@@ -58,6 +60,33 @@ type QueryExecutor struct {
 	statMap *expvar.Map
 }
 
+// ExecutionContext contains state that the query is currently executing with.
+type ExecutionContext struct {
+	// The statement ID of the executing query.
+	StatementID int
+
+	// Output channel where results and errors should be sent.
+	Results chan *influxql.Result
+
+	// The database the query is running against.
+	Database string
+
+	// The requested maximum number of points to return in each result.
+	ChunkSize int
+
+	// A channel that is closed when the query is interrupted.
+	InterruptCh <-chan struct{}
+}
+
+// PluginExecutor allows a statement to be intercepted and executed by a custom
+// provider rather than the QueryExecutor.
+type PluginExecutor interface {
+	// ExecuteStatement attempts to execute a statement using a plugin.
+	// If the plugin executes the statement, it should return true.
+	// Results and errors should be sent over the results channel.
+	ExecuteStatement(stmt influxql.Statement, ctx *ExecutionContext) bool
+}
```

Comment:
Yea. I wasn't certain which name to use at the time, but I think `StatementExecutor` is good.